### PR TITLE
Pullupstream/v0.5.0 rc4

### PR DESCRIPTION
**Pull request by arnaubennassar**
Instead of reviewing this PR code diff, which is absolutely insane, review the code diff vs zkevm-node. Instructions on `docs/diff/diff.md` included on this branch

### CODE REVIEW

File: docs/diff/diff.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Diff
+
+This repo is a fork from [zkevm-node](https://github.com/0xPolygonHermez.zkevm-node). The puirpose of the fork is to implement tha Validium consensus, enabling data availability to be posted outside of L1.
```
Comment (Stefan-Ethernal) [inline docs/diff/diff.md]: Minor typos:

```suggestion
This repo is a fork from [zkevm-node](https://github.com/0xPolygonHermez.zkevm-node). The purpose of the fork is to implement the Validium consensus, enabling data availability to be posted outside of L1.
```

File: dataavailability/dataavailability.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+	"github.com/0xPolygonHermez/zkevm-node/etherman/types"
+	"github.com/0xPolygonHermez/zkevm-node/log"
+	"github.com/0xPolygonHermez/zkevm-node/state"
+	"github.com/0xPolygonHermez/zkevm-node/synchronizer/common/syncinterfaces"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// DataAvailability implements an abstract data avaiolability integration
```
Comment (Stefan-Ethernal) [inline dataavailability/dataavailability.go]: Minor typo:

```suggestion
// DataAvailability implements an abstract data availability integration
```

File: dataavailability/dataavailability.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	}
+	actualTransactionsHash := crypto.Keccak256Hash(transactionsData)
+	if !found || expectedTransactionsHash != actualTransactionsHash {
+		if found {
+			log.Warnf(unexpectedHashTemplate, batchNum, expectedTransactionsHash, actualTransactionsHash)
+		}
+
+		if !d.isTrustedSequencer {
+			log.Info("trying to get data from trusted sequencer")
+			data, err := d.getDataFromTrustedSequencer(batchNum, expectedTransactionsHash)
```
Comment (Stefan-Ethernal) [inline dataavailability/dataavailability.go:81]: Sorry for the ignorance, but is this logic inverted? In case `isTrustedSequencer == false`, we are fetching data from the trusted sequencer. It looks a bit strange on the first glance, but otoh it might be that if the "local sequencer" is not a trusted one, then retrieve data from the trusted one, right?

Anyways, probably an inline comment would help in understanding this part better.

File: dataavailability/dataavailability.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		return data, nil
+	}
+	return transactionsData, nil
+}
+
+const unexpectedHashTemplate = "missmatch on transaction data for batch num %d. Expected hash %s, actual hash: %s"
+
+func (d *DataAvailability) getDataFromTrustedSequencer(batchNum uint64, expectedTransactionsHash common.Hash) ([]byte, error) {
+	b, err := d.zkEVMClient.BatchByNumber(d.ctx, big.NewInt(int64(batchNum)))
+	if err != nil {
```
Comment (Stefan-Ethernal) [inline dataavailability/dataavailability.go:106]: Is it safe to do down-casting from `uint64` to `int64` here? I'd suggest safer option, wdyt?

```suggestion
	b, err := d.zkEVMClient.BatchByNumber(d.ctx, new(big.Int).SetUint64(batchNum))
```

File: dataavailability/datacommittee/datacommittee.go [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+type DataCommittee struct {
+	AddressesHash      common.Hash
+	Members            []DataCommitteeMember
+	RequiredSignatures uint64
+}
+
+// DataCommitteeBackend implements the DAC integration
+type DataCommitteeBackend struct {
+	dataCommitteeContract      *polygondatacommittee.Polygondatacommittee
+	l2Coinbase                 common.Address
```
Comment (Stefan-Ethernal) [inline dataavailability/datacommittee/datacommittee.go]: It is currently unused, are we going to need it in the future, or we can remove it at this point?

File: test/scripts/sendForcedBatch/main.go [hunk 6]
```diff
@@ -105,34 +119,34 @@ func sendForcedBatches(cliCtx *cli.Context) error {
...
+	// 		log.Error("error sending activateForceBatches. Error: ", err)
+	// 		return err
+	// 	}
+	// 	err = operations.WaitTxToBeMined(ctx, ethClient, tx, operations.DefaultTimeoutTxToBeMined)
+	// 	if err != nil {
+
+	// 		log.Error("error waiting tx to be mined. Error: ", err)
+	// 		return err
+	// 	}
+	// }
```
Comment (Stefan-Ethernal) [inline test/scripts/sendForcedBatch/main.go]: Is this going to be needed in the future perhaps?

File: etherman/interfaces.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package etherman
+
+import "github.com/ethereum/go-ethereum/common"
+
+type dataAvailabilitier interface {
```
Comment (Stefan-Ethernal) [inline etherman/interfaces.go]: `dataAvailabilitier` is hard to spell and even more to pronounce in verbal comunication, that's why I'm suggesting this change if it is still semantically correct :slightly_smiling_face: WDYT?

```suggestion
type l2DataProvider interface {
```

File: etherman/etherman.go [hunk 8]
```diff
@@ -782,16 +1207,100 @@ func (etherMan *Client) sequencedBatchesEvent(ctx context.Context, vLog types.Lo
...
+		var sequencesValidium []polygonzkevm.PolygonValidiumEtrogValidiumBatchData
+		err = json.Unmarshal(bytedata, &sequencesValidium)
+		if err != nil {
+			return nil, err
+		}
+		coinbase := (data[1]).(common.Address)
+		sequencedBatches := make([]SequencedBatch, len(sequencesValidium))
+		for i, seq := range sequencesValidium {
+			bn := lastBatchNumber - uint64(len(sequencesValidium)-(i+1))
+			batchL2Data, err := da.GetBatchL2Data(bn, sequencesValidium[i].TransactionsHash)
```
Comment (Stefan-Ethernal) [inline etherman/etherman.go:1281]: `da` field might be `nil`, so wdyt about following null object pattern (providing dummy implementation for `nil` case) and not even care if reference is undefined or not. WDYT?

File: dataavailability/interfaces.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+// DABackender is the interface for a DA backend
+type DABackender interface {
+	// Init initializes the DABackend
+	Init() error
+	// GetData retrieve the data of a batch from the DA backend. The returned data must be the pre-image of the hash
+	GetData(batchNum uint64, hash common.Hash) ([]byte, error)
+	// PostSequence sends the sequence data to the data availability backend, and returns the dataAvailabilityMessage
+	// as expected by the contract
+	PostSequence(ctx context.Context, batchesData [][]byte) ([]byte, error)
+}
```
Comment (Stefan-Ethernal) [inline dataavailability/interfaces.go]: WDYT about such interface segregation, because if I'm not mistaken, it seems that sometimes you need to implement just the `BatchDataProvider` (reading batches) and sometimes only `SequenceSender` (posting batches, e.g. here https://github.com/0xPolygon/cdk-validium-node/blob/9d186be306740f2fa9e1dcc4186124baa65c6d77/sequencesender/interfaces.go#L42). Does it make sense, or it is an overkill?

```suggestion
// BatchDataProvider is used to retrieve batch data
type BatchDataProvider interface {
	// GetData retrieve the data of a batch from the DA backend. The returned data must be the pre-image of the hash
	GetData(batchNum uint64, hash common.Hash) ([]byte, error)
}

// SequenceSender is used to send provided sequence of batches
type SequenceSender interface {
	// PostSequence sends the sequence data to the data availability backend, and returns the dataAvailabilityMessage
	// as expected by the contract
	PostSequence(ctx context.Context, batchesData [][]byte) ([]byte, error)
}

type DABackender interface {
	BatchDataProvider
	SequenceSender
	// Init initializes the DABackend
	Init() error
}
```

File: cmd/run.go [hunk 10]
```diff
@@ -278,12 +288,74 @@ func runMigrations(c db.Config, name string) {
...
+		daBackend, err = datacommittee.New(
+			c.Etherman.URL,
+			dacAddr,
+			c.SequenceSender.L2Coinbase,
+			pk,
+			&dataCommitteeClient.Factory{},
+		)
+		if err != nil {
+			return nil, err
+		}
```
Comment (Stefan-Ethernal) [inline cmd/run.go:356]: Shall we add default statement and return an error perhaps in order to fail fast, because it would mean we have retrieved some unknown DA protocol name?

File: cmd/run.go [hunk 11]
```diff
@@ -278,12 +288,74 @@ func runMigrations(c db.Config, name string) {
...
-	return etherman, nil
+	da, err := newDataAvailability(c, st, ethman, false)
+	if err != nil {
+		return nil, err
+	}
+	return etherman.NewClient(c.Etherman, c.NetworkConfig.L1Config, da)
+}
+
+func newDataAvailability(c config.Config, st *state.State, etherman *etherman.Client, isSequenceSender bool) (*dataavailability.DataAvailability, error) {
+	var trustedSequencerURL string
```
Comment (goran-ethernal) [inline cmd/run.go]:
```suggestion
	var (
	     trustedSequencerURL string
	     err error
	 )
```

File: cmd/run.go [hunk 12]
```diff
@@ -278,12 +288,74 @@ func runMigrations(c db.Config, name string) {
...
+	var err error
+	if !c.IsTrustedSequencer {
+		if c.Synchronizer.TrustedSequencerURL != "" {
+			trustedSequencerURL = c.Synchronizer.TrustedSequencerURL
+		} else {
+			log.Debug("getting trusted sequencer URL from smc")
+			trustedSequencerURL, err = etherman.GetTrustedSequencerURL()
+			if err != nil {
+				log.Fatal("error getting trusted sequencer URI. Error: %v", err)
```
Comment (goran-ethernal) [inline cmd/run.go]: I think it's ok to just return an error if it happens in this function, just to be consistent, since on some places in function we do `log.Fatal`, and on some, we return an error. This is called on `newEtherman` function, and on `etherman, err := newEtherman(*c, st)` if an error happens, it does `log.Fatal`. Same happens on `createSequenceSender`.

File: cmd/run.go [hunk 13]
```diff
@@ -278,12 +288,74 @@ func runMigrations(c db.Config, name string) {
...
+
+	// Backend specific config
+	daProtocolName, err := etherman.GetDAProtocolName()
+	if err != nil {
+		log.Fatal("error getting data availability protocol name: %v", err)
+	}
+	var daBackend dataavailability.DABackender
+	switch daProtocolName {
+	case string(dataavailability.DataAvailabilityCommittee):
+		var pk *ecdsa.PrivateKey
```
Comment (goran-ethernal) [inline cmd/run.go]:
```suggestion
		var (
		    pk *ecdsa.PrivateKey
		    err error
		)
```

File: cmd/run.go [hunk 14]
```diff
@@ -492,8 +563,20 @@ func newState(ctx context.Context, c *config.Config, l2ChainID uint64, forkIDInt
 		MaxLogsBlockRange:            c.RPC.MaxLogsBlockRange,
 		MaxNativeBlockHashBlockRange: c.RPC.MaxNativeBlockHashBlockRange,
 	}
+	allLeaves, err := stateDb.GetAllL1InfoRootEntries(ctx, nil)
+	if err != nil {
+		log.Fatal("error getting all leaves. Error: ", err)
+	}
+	var leaves [][32]byte
```
Comment (goran-ethernal) [inline cmd/run.go:578]:
```suggestion
	leaves := make([][32]byte, 0, len(allLeaves))
```

File: dataavailability/dataavailability.go [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+	state       stateInterface
+	zkEVMClient syncinterfaces.ZKEVMClientInterface
+	backend     DABackender
+
+	ctx context.Context
+}
+
+// New creates a DataAvailability instance
+func New(
+	IsTrustedSequencer bool,
```
Comment (goran-ethernal) [inline dataavailability/dataavailability.go]:
```suggestion
	IsTrustedSequencer bool,
```
```suggestion
	isTrustedSequencer bool,
```

File: cmd/run.go [hunk 16]
```diff
@@ -278,12 +288,74 @@ func runMigrations(c db.Config, name string) {
...
+	if err != nil {
+		return nil, err
+	}
+	return etherman.NewClient(c.Etherman, c.NetworkConfig.L1Config, da)
+}
+
+func newDataAvailability(c config.Config, st *state.State, etherman *etherman.Client, isSequenceSender bool) (*dataavailability.DataAvailability, error) {
+	var trustedSequencerURL string
+	var err error
+	if !c.IsTrustedSequencer {
```
Comment (Stefan-Ethernal) [inline cmd/run.go:311]: Nitpick:
```suggestion
	var (
		trustedSequencerURL string
		err                 error
	)
```

File: cmd/run.go [hunk 17]
```diff
@@ -278,12 +288,74 @@ func runMigrations(c db.Config, name string) {
...
+		if c.Synchronizer.TrustedSequencerURL != "" {
+			trustedSequencerURL = c.Synchronizer.TrustedSequencerURL
+		} else {
+			log.Debug("getting trusted sequencer URL from smc")
```
Comment (Stefan-Ethernal) [inline cmd/run.go:315]: Just curious, what smc abbreviation stands for, is it smart contract?

File: cmd/run.go [hunk 18]
```diff
@@ -278,12 +288,74 @@ func runMigrations(c db.Config, name string) {
...
+	// Backend specific config
+	daProtocolName, err := etherman.GetDAProtocolName()
+	if err != nil {
+		log.Fatal("error getting data availability protocol name: %v", err)
+	}
+	var daBackend dataavailability.DABackender
+	switch daProtocolName {
+	case string(dataavailability.DataAvailabilityCommittee):
+		var pk *ecdsa.PrivateKey
+		var err error
```
Comment (Stefan-Ethernal) [inline cmd/run.go]:
```suggestion
		var (
            pk *ecdsa.PrivateKey
		    err error
        )
```

File: dataavailability/dataavailability.go [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+	if err != nil {
+		if err == state.ErrNotFound {
+			found = false
+		} else {
+			return nil, fmt.Errorf("failed to get batch data from state for batch num %d: %w", batchNum, err)
+		}
+	}
+	actualTransactionsHash := crypto.Keccak256Hash(transactionsData)
+	if !found || expectedTransactionsHash != actualTransactionsHash {
+		if found {
```
Comment (goran-ethernal) [inline dataavailability/dataavailability.go:75]: Should we log the `not found` case as well?

File: dataavailability/dataavailability.go [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+			log.Warnf(unexpectedHashTemplate, batchNum, expectedTransactionsHash, actualTransactionsHash)
+		}
+
+		if !d.isTrustedSequencer {
+			log.Info("trying to get data from trusted sequencer")
+			data, err := d.getDataFromTrustedSequencer(batchNum, expectedTransactionsHash)
+			if err != nil {
+				log.Error(err)
```
Comment (goran-ethernal) [inline dataavailability/dataavailability.go]:
```suggestion
				log.Error("failed to get data from trusted sequencer: %w", err)
```

File: dataavailability/dataavailability.go [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+			} else {
+				return data, nil
+			}
+		}
+
+		log.Info("trying to get data from the data availability backend")
+		data, err := d.backend.GetData(batchNum, expectedTransactionsHash)
+		if err != nil {
+			log.Error(err)
```
Comment (goran-ethernal) [inline dataavailability/dataavailability.go]:
```suggestion
			log.Error(err)
```
```suggestion
			log.Error("failed to get data from the data availability backend: %w", err)
```

File: state/pgstatestorage/pgstatestorage.go [hunk 22]
```diff
@@ -0,0 +1,370 @@
...
+			return nil, err
+		}
+		nativeBlockHashes = append(nativeBlockHashes, common.HexToHash(nativeBlockHash))
+	}
+	return nativeBlockHashes, nil
+}
+
+// GetBatchL2DataByNumber returns the batch L2 data of the given batch number.
+func (p *PostgresStorage) GetBatchL2DataByNumber(ctx context.Context, batchNumber uint64, dbTx pgx.Tx) ([]byte, error) {
+	getBatchL2DataByBatchNumber := "SELECT raw_txs_data FROM state.batch WHERE batch_num = $1"
```
Comment (Stefan-Ethernal) [inline state/pgstatestorage/pgstatestorage.go]: Does it make sense to extract this into constant?

File: dataavailability/dataavailability.go [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+			} else {
+				return nil, fmt.Errorf("data not found on the local DB, nor from the trusted sequencer nor on any data committee member")
+			}
+		}
+		return data, nil
+	}
+	return transactionsData, nil
+}
+
+const unexpectedHashTemplate = "missmatch on transaction data for batch num %d. Expected hash %s, actual hash: %s"
```
Comment (goran-ethernal) [inline dataavailability/dataavailability.go]: I would move global constants and variables to the top of the file.

File: dataavailability/dataavailability.go [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+
+// New creates a DataAvailability instance
+func New(
+	IsTrustedSequencer bool,
+	backend DABackender,
+	state stateInterface,
+	zkEVMClient syncinterfaces.ZKEVMClientInterface,
+) (*DataAvailability, error) {
+	da := &DataAvailability{
+		isTrustedSequencer: IsTrustedSequencer,
```
Comment (Stefan-Ethernal) [inline dataavailability/dataavailability.go]:
```suggestion
	isTrustedSequencer bool,
	backend DABackender,
	state stateInterface,
	zkEVMClient syncinterfaces.ZKEVMClientInterface,
) (*DataAvailability, error) {
	da := &DataAvailability{
		isTrustedSequencer: isTrustedSequencer,
```

File: dataavailability/datacommittee/datacommittee.go [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+// getCurrentDataCommitteeMembers return the currently registered data committee members
+func (d *DataCommitteeBackend) getCurrentDataCommitteeMembers() ([]DataCommitteeMember, error) {
+	members := []DataCommitteeMember{}
+	nMembers, err := d.dataCommitteeContract.GetAmountOfMembers(&bind.CallOpts{Pending: false})
+	if err != nil {
+		return nil, fmt.Errorf("error getting GetAmountOfMembers from L1 SC: %w", err)
+	}
+	for i := int64(0); i < nMembers.Int64(); i++ {
```
Comment (Stefan-Ethernal) [inline dataavailability/datacommittee/datacommittee.go]: Minor optimization, providing initial capacity to the slice WDYT?

```suggestion
	nMembers, err := d.dataCommitteeContract.GetAmountOfMembers(&bind.CallOpts{Pending: false})
	if err != nil {
		return nil, fmt.Errorf("error getting GetAmountOfMembers from L1 SC: %w", err)
	}
	members := make([]DataCommitteeMember, 0, nMembers.Int64())
```

File: dataavailability/datacommittee/datacommittee.go [hunk 26]
```diff
@@ -0,0 +1,292 @@
...
+		d.committeeMembers = committee.Members
+		if len(committee.Members) > 0 {
+			selectedCommitteeMember = rand.Intn(len(committee.Members)) //nolint:gosec
+		}
+	}
+	d.selectedCommitteeMember = selectedCommitteeMember
+	return nil
+}
+
+const unexpectedHashTemplate = "missmatch on transaction data for batch num %d. Expected hash %s, actual hash: %s"
```
Comment (goran-ethernal) [inline dataavailability/datacommittee/datacommittee.go]: Move to the top of the file.

**Review by Stefan-Ethernal — APPROVED** [review]
Linter is complaining that there are some unused fields and functions in `cmd/policy.go`, although they are indeed used. Maybe we should try to suppress `unused` warning for that file?

Leaving some comments to consider.

**Review by Stefan-Ethernal — APPROVED** [review]
Linter is complaining that there are some unused fields and functions in `cmd/policy.go`, although they are indeed used. Maybe we should try to suppress `unused` warning for that file?

Leaving some comments to consider.

File: dataavailability/datacommittee/datacommittee.go [hunk 27]
```diff
@@ -0,0 +1,292 @@
...
+
+// GetData returns the data from the DAC. It checks that it matches with the expected hash
+func (d *DataCommitteeBackend) GetData(batchNum uint64, hash common.Hash) ([]byte, error) {
```
Comment (goran-ethernal) [inline dataavailability/datacommittee/datacommittee.go]: By looking at this `for` loop, seems there can be cases where it can run forever. Should we add some timeout to stop its execution after some time?

File: dataavailability/datacommittee/datacommittee.go [hunk 28]
```diff
@@ -0,0 +1,292 @@
...
+	// Request signatures to all members in parallel
+	ch := make(chan signatureMsg, len(committee.Members))
+	signatureCtx, cancelSignatureCollection := context.WithCancel(ctx)
+	for _, member := range committee.Members {
+		go requestSignatureFromMember(signatureCtx, *signedSequence, member, ch)
+	}
+
+	// Collect signatures
+	msgs := []signatureMsg{}
+	var collectedSignatures uint64
```
Comment (goran-ethernal) [inline dataavailability/datacommittee/datacommittee.go]:
```suggestion
	var (
	    collectedSignatures uint64
	    failedToCollect uint64
	 )
```

File: dataavailability/datacommittee/datacommittee.go [hunk 29]
```diff
@@ -0,0 +1,292 @@
...
+		return
+	}
+	ch <- signatureMsg{
+		addr:      member.Addr,
+		signature: signature,
+	}
+}
+
+func buildSignaturesAndAddrs(msgs signatureMsgs, members []DataCommitteeMember) []byte {
+	res := []byte{}
```
Comment (goran-ethernal) [inline dataavailability/datacommittee/datacommittee.go]:
```suggestion
	res := make([]byte, 0, len(msgs)+len(members))
```

File: dataavailability/datacommittee/datacommittee_test.go [hunk 30]
```diff
@@ -0,0 +1,127 @@
...
+
+func init() {
+	log.Init(log.Config{
+		Level:   "debug",
+		Outputs: []string{"stderr"},
+	})
+}
+
+// This function prepare the blockchain, the wallet with funds and deploy the smc
+func newTestingEnv() (
```
Comment (goran-ethernal) [inline dataavailability/datacommittee/datacommittee_test.go]: Pass `testing.T` to it, and set `t.Helper()` at the beginning of the function to mark it as test helper function.

File: dataavailability/datacommittee/datacommittee_test.go [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+	dac, ethBackend, da, err = newSimulatedDacman(auth)
+	if err != nil {
+		log.Fatal(err)
+	}
+	return dac, ethBackend, auth, da
+}
+
+// NewSimulatedEtherman creates an etherman that uses a simulated blockchain. It's important to notice that the ChainID of the auth
+// must be 1337. The address that holds the auth will have an initial balance of 10 ETH
+func newSimulatedDacman(auth *bind.TransactOpts) (
```
Comment (goran-ethernal) [inline dataavailability/datacommittee/datacommittee_test.go]: Pass `testing.T` to it, and set `t.Helper()` at the beginning of the function to mark it as test helper function.

File: etherman/etherman.go [hunk 32]
```diff
@@ -467,24 +754,82 @@ func (etherMan *Client) updateZkevmVersion(ctx context.Context, vLog types.Log,
...
 }
 
+func (etherMan *Client) updateL1InfoTreeEvent(ctx context.Context, vLog types.Log, blocks *[]Block, blocksOrder *map[common.Hash][]Order) error {
+	log.Debug("UpdateL1InfoTree event detected")
+	globalExitRootL1InfoTree, err := etherMan.GlobalExitRootManager.ParseUpdateL1InfoTree(vLog)
+	if err != nil {
+		return err
+	}
+
+	var gExitRoot GlobalExitRoot
```
Comment (goran-ethernal) [inline etherman/etherman.go:774]:
```suggestion
	 gExitRoot := GlobalExitRoot {
	      MainnetExitRoot: globalExitRootL1InfoTree.MainnetExitRoot,
	      RollupExitRoot: globalExitRootL1InfoTree.RollupExitRoot,
	      BlockNumber: vLog.BlockNumber,
	      GlobalExitRoot: hash(globalExitRootL1InfoTree.MainnetExitRoot, globalExitRootL1InfoTree.RollupExitRoot),
	 }
```

File: etherman/etherman_test.go [hunk 33]
```diff
@@ -193,21 +178,31 @@ func TestSequencedBatchesEvent(t *testing.T) {
...
-	assert.Equal(t, 4, len(blocks))
-	assert.Equal(t, 1, len(blocks[3].SequencedBatches))
-	assert.Equal(t, crypto.Keccak256Hash(common.Hex2Bytes(rawTxs)), common.Hash(blocks[3].SequencedBatches[0][1].TransactionsHash))
-	assert.Equal(t, currentBlock.Time(), blocks[3].SequencedBatches[0][0].Timestamp)
-	assert.Equal(t, ger, blocks[3].SequencedBatches[0][0].GlobalExitRoot)
-	assert.Equal(t, auth.From, blocks[3].SequencedBatches[0][0].Coinbase)
-	assert.Equal(t, auth.From, blocks[3].SequencedBatches[0][0].SequencerAddr)
-	assert.Equal(t, currentBlock.Time(), blocks[3].SequencedBatches[0][0].MinForcedTimestamp)
-	assert.Equal(t, 0, order[blocks[3].BlockHash][0].Pos)
+	// t.Log("Blocks: ", blocks)
```
Comment (goran-ethernal) [inline etherman/etherman_test.go]: Should we remove this commented code?

File: dataavailability/dataavailability.go (hunk 3 above)
Comment (arnaubennassar) [reply dataavailability/dataavailability.go:81]: isTrustedSequencer is a config param widely used in the app, kind of a core concept. Basically is saying "I'm the trusted sequencer? yes/no"

It doesnt make sense to pull DA from trusted sequencer if I'm the trusted sequecer already

File: test/scripts/sendForcedBatch/main.go (hunk 6 above)
Comment (arnaubennassar) [reply test/scripts/sendForcedBatch/main.go]: yes, I've diverged from the upstream here due to being ahead in contract integration. IMO it felt better to leave this commented rather than remove test 😅 . But this is going to be fixed in next batch of pulled changed

File: etherman/interfaces.go (hunk 7 above)
Comment (arnaubennassar) [reply etherman/interfaces.go]: agree in that the name isnt great, but since DA is kind of a buzz word in the industry, I think we should follow it. But I can change it to `dataAvailabilityProvider`

File: etherman/etherman.go (hunk 8 above)
Comment (arnaubennassar) [reply etherman/etherman.go:1281]: mmmmm, in that context, da should never be nil. In fact the possibility for it to be nil is there cuz of being a fork from something that doesnt need it. I'm ok having a good old panic here if for whatever crazy reason this happens to be nil

File: test/scripts/sendForcedBatch/main.go [hunk 34]
```diff
@@ -105,34 +119,34 @@ func sendForcedBatches(cliCtx *cli.Context) error {
...
-			return err
-		}
-		err = operations.WaitTxToBeMined(ctx, ethClient, tx, operations.DefaultTimeoutTxToBeMined)
-		if err != nil {
-
-			log.Error("error waiting tx to be mined. Error: ", err)
-			return err
-		}
-	}
+	// // Allow forced batches in smart contract if disallowed
```
Comment (vcastellm) [inline test/scripts/sendForcedBatch/main.go]: Is this a leftover?

File: cmd/run.go (hunk 17 above)
Comment (arnaubennassar) [reply cmd/run.go:315]: it's just smart contract. IMO SC is way better, but this is inherited from upstream

File: state/pgstatestorage/pgstatestorage.go (hunk 22 above)
Comment (arnaubennassar) [reply state/pgstatestorage/pgstatestorage.go]: modified to be a constant. But wee decided to put the queries inside each function. Before we used to have a massive `const` statement with all the sql, and it was painful for the queries and the functions that called the queries to be so separated in space. There is almost an exact 1:1 relation between query and function

File: cmd/run.go (hunk 14 above)
Comment (arnaubennassar) [reply cmd/run.go:578]: this is outside of the diff vs zkevm-node

File: dataavailability/dataavailability.go (hunk 19 above)
Comment (arnaubennassar) [reply dataavailability/dataavailability.go:75]: the reasond behind logging only if found, is because that means that there is a hash missmatch. This implies either a bug or some actor trying to do nasty stuff. But not having found something is expected (for instance when syncing a node from scratch, that log would appear basically once per batch without adding any value)

File: dataavailability/datacommittee/datacommittee.go (hunk 27 above)
Comment (arnaubennassar) [reply dataavailability/datacommittee/datacommittee.go]: the code is hard to follow, but should always break after reaching the initial committee member that has been queried

File: etherman/etherman.go (hunk 32 above)
Comment (arnaubennassar) [reply etherman/etherman.go:774]: that comes from upstream, I'm consciously choosing to reduce code diff over clean code

File: etherman/etherman_test.go (hunk 33 above)
Comment (arnaubennassar) [reply etherman/etherman_test.go]: comes from upstream

File: test/scripts/sendForcedBatch/main.go (hunk 34 above)
Comment (arnaubennassar) [reply test/scripts/sendForcedBatch/main.go:128]: comes from upstream

Comment (arnaubennassar) [reply test/scripts/sendForcedBatch/main.go]: I had to upgrade contracts beyond the version that was used in the commit I'm pulling changes from upstream. This made me comment a bunch of code related to forced batches. In the next pulling this code should be uncommented with some chenges

File: dataavailability/dataavailability.go [hunk 35]
```diff
@@ -0,0 +1,117 @@
...
+	if !found || expectedTransactionsHash != actualTransactionsHash {
+		if found {
+			log.Warnf(unexpectedHashTemplate, batchNum, expectedTransactionsHash, actualTransactionsHash)
+		}
+
+		if !d.isTrustedSequencer {
+			log.Info("trying to get data from trusted sequencer")
+			data, err := d.getDataFromTrustedSequencer(batchNum, expectedTransactionsHash)
+			if err != nil {
+				log.Error("failed to get data from trusted sequencer: %w", err)
```
Comment (vcastellm) [inline dataavailability/dataavailability.go]: Are these really errors or warnings?

Comment (arnaubennassar) [reply dataavailability/dataavailability.go]: will lower log level in next PR

File: sequencesender/sequencesender.go [hunk 36]
```diff
@@ -35,18 +34,18 @@ type SequenceSender struct {
 	ethTxManager ethTxManager
 	etherman     etherman
 	eventLog     *event.EventLog
-	privKey      *ecdsa.PrivateKey
+	da           dataAbilitier
```
Comment (vcastellm) [inline sequencesender/sequencesender.go:37]: `dataEnabler` better?

**Review by vcastellm — APPROVED** [review]
Aside from minor comments  LGTM

**Review by goran-ethernal — APPROVED** [review]
